drivers/sensor/adt7310/adt7310.c

```diff
@@ -0,0 +1,303 @@
+	int ret;
+	const struct spi_buf tx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf rx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf_set tx = { .buffers = &tx_buf, .count = 1 };
+	const struct spi_buf_set rx = { .buffers = &rx_buf, .count = 1 };
+
+	ret = spi_transceive_dt(&cfg->bus, &tx, &rx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	memcpy(val, cmd_buf + 1, 2);
+	*val = sys_be16_to_cpu(*val);
+
+	return 0;
+}
+
+static int adt7310_temp_reg_write(const struct device *dev, uint8_t reg, int16_t val)
+{
+	const struct adt7310_dev_config *cfg = dev->config;
+	uint8_t cmd_buf[3];
+	const struct spi_buf tx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf rx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf_set tx = { .buffers = &tx_buf, .count = 1 };
+	const struct spi_buf_set rx = { .buffers = &rx_buf, .count = 1 };
+
+	cmd_buf[0] = ADT7310_WRITE_CMD | (reg << 3);
+	val = sys_cpu_to_be16(val);
+	memcpy(&cmd_buf[1], &val, sizeof(val));
+
+	return spi_transceive_dt(&cfg->bus, &tx, &rx);
+}
+
+static int adt7310_reg_read(const struct device *dev, uint8_t reg, uint8_t *val)
+{
+	const struct adt7310_dev_config *cfg = dev->config;
+	uint8_t cmd_buf[2] = { ADT7310_READ_CMD | (reg << 3) };
+	const struct spi_buf tx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf rx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf_set tx = { .buffers = &tx_buf, .count = 1 };
+	const struct spi_buf_set rx = { .buffers = &rx_buf, .count = 1 };
+	int ret;
+
+	ret = spi_transceive_dt(&cfg->bus, &tx, &rx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*val = cmd_buf[1];
+
+	return 0;
+}
+
+static int adt7310_reg_write(const struct device *dev, uint8_t reg, uint8_t val)
+{
+	const struct adt7310_dev_config *cfg = dev->config;
+	uint8_t cmd_buf[2] = { ADT7310_WRITE_CMD | (reg << 3), val };
+	const struct spi_buf tx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf rx_buf = { .buf = cmd_buf, .len = sizeof(cmd_buf) };
+	const struct spi_buf_set tx = { .buffers = &tx_buf, .count = 1 };
+	const struct spi_buf_set rx = { .buffers = &rx_buf, .count = 1 };
+
+	return spi_transceive_dt(&cfg->bus, &tx, &rx);
+}
+
+static int adt7310_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct adt7310_data *drv_data = dev->data;
+	int16_t value;
+	int ret;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		return -ENOTSUP;
+	}
+
+	ret = adt7310_temp_reg_read(dev, ADT7310_REG_TEMP, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* The quantization step size is 0.0078125. This would imply a multiplication */
+	/* by 7812.5 when converting to micro degrees in adt7310_channel_get(). Instead */
+	/* drop the lsb, which doubles the quantization step. The float multiplication */
+	/* is therefore avoided by multiplying with 15625. */
+	drv_data->sample = value >> 1;
```

Comment:
You can multiply before the shift to avoid throwing away a bit of precision. I suggest removing this section and modifying `adt7310_channel_get` to:
```c
value = (int32_t)drv_data->sample * ADT7310_TEMP_SCALE >> 1;
```

Reply:
Good idea, I updated the PR. 
We should also then probably update drivers/sensor/adt7420
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/adt7420/adt7420.c#L124